README.md

# SDEtools

<!-- badges: start -->
<!-- badges: end -->

SDEtools provides functions for analyzing models based on stochastic differential equations (SDE's). These analysis include simulation of sample paths, solving the Kolmogorov equations that govern transition probabilities, state estimation, and optimal control.

## Installation

You can install the development version of SDEtools from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("Uffe-H-Thygesen/SDEtools")
```

## Example

This is a basic example which shows you how to simulate a sample path, and compute the transition probability, for a double well model:

``` r
library(SDEtools)
## basic example code
f <- function(x) x-x^3
g <- function(x) 0.5

tv <- seq(0,6,0.01)
sim <- heun(f,g,tv,0)

plot(tv,sim$X,type="l",ylim=c(-1.5,1.5))

xi <- seq(-1.5,1.5,0.01)
xc <- cell.centers(xi)

D <- function(x) 0.5*g(x)^2
G <- fvade(f,D,xi,'r')

rho <- StationaryDistribution(G)

lines(200*rho,xc)
```



NEWS.md
# SDEtools 0.1

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(SDEtools)

test_check("SDEtools")


tests/testthat/test_solvers.R
library(SDEtools)

test_that("SDE simulation: Dimensions of states and noise channels",{
  times <- 0:10

  nx <- 3
  f <- function(t,x) -x
  g <- function(t,x) x
  sol <- euler(f,g,times,rep(1,nx))

  expect_equal(dim(sol$X),c(length(times),nx))

  nb <- 2
  g <- function(t,x) matrix(runif(nx*nb),ncol=nb,nrow=nx)
  sol <- euler(f,g,times,rep(1,nx))

  expect_equal(dim(sol$X),c(length(times),nx))
})

test_that("Linear SDEs",{
  sol <- dLinSDE(-1,1,1,S0=0.5)
  expect_equal(as.numeric(sol$eAt),exp(-1))
  expect_equal(as.numeric(sol$St),0.5)

  A <- matrix(c(0,-1,1,0),nrow=2,ncol=2)
  G <- diag(c(1,1))
  sol <- dLinSDE(A,G,2*pi,x0=c(1,0))
  expect_equal(as.numeric(sol$EX),c(1,0))
  expect_equal(sol$St,diag(rep(2*pi,2)))

  sol <- dLinSDE(-1,1,2,x0=1,u=1)
  expect_equal(as.numeric(sol$EX),1)

  sol <- dLinSDE(0,1,2,x0=1,u=matrix(c(1,3),nrow=1,ncol=2))
  expect_equal(as.numeric(sol$EX),5)
})


R/control.R
Riccati <- function(time,Ss,par)
{
  S <- matrix(Ss[1:(length(Ss)-1)],nrow=nrow(par$A))
  SA <- S %*% par$A
  dSdt <- SA + t(SA) - S %*% par$FRiF %*% S + par$Q
  dsdt <- 0.5*sum(diag(par$Gt %*% S %*% par$G))

  return(list(c(as.numeric(dSdt),dsdt)))
}

#' Solve the time-varying LQR (Linear Quadratic Regulator) problem
#'
#' @param  times numeric vector of time points
#' @param  A system matrix, n-by-n numeric array
#' @param  F control matrix, n-by-m numeric array
#' @param  G noise matrix, n-by-l numeric array
#' @param  Q state penalty running cost, n-by-n numeric array specifying the quadratic form in x in the running cost
#' @param  R control penalty in running cost, m-by-m numeric array specifying the quadratic form in u in the running cost
#' @param  P terminal state penalty, n-by-n numeric array specifying the quadratic form in x in the terminal cost
#' @return A list containing
#' times, A, F, G, Q, R, P: The input arguments
#' S the value function. A length(times)*n*n array containing, for each time point, the quadratic form in x in the value function
#' s the value function. A length(times) vector containing, for each time point, the off-set in the value function, i.e. the value for x=0
#' L the optimal control. A length(times)*m*n array containing, for each time point, the gain matrix Lt that maps states to optimal controls
#'
#' @examples
#' # Feedback control of a harmonic oscillator
#' A <- array(c(0,1,-1,0),c(2,2))
#' F <- G <- array(c(0,1),c(2,1))
#' Q <- diag(rep(1,2))
#' R <- 1
#' times <- seq(0,10,0.1)
#' sol <- lqr(times,A,F,G,Q,R)
#' matplot(-times,array(sol$S,c(length(times),length(A))),type="l",xlab="Time",ylab="Elements in S")
#' legend("topright",c("xx","xv","vx","vv"),lty=c("solid","dashed","dashed","dotted"),col=1:4)
#' matplot(-times,array(sol$L,c(length(times),length(F))),type="l",
#'          xlab="Time",ylab="Feedback control",lty=c("solid","dashed"))
#' legend("topright",c("Position","Velocity"),lty=c("solid","dashed"),col=1:2)
#' @export
lqr <- function(times,A,F,G,Q,R,P=NULL)
{
  RiF <- solve(R,t(F))

  if(is.null(P)) P <- array(0,rep(nrow(A),2))

  par <- list(A = A,G = G,Gt = t(G),FRiF = F %*% RiF,Q=Q)

  require(deSolve)

  sol <- ode(y = c(as.numeric(P),0),times=times,func=Riccati,parms=par)

  SvecToGain <- function(Ss) as.numeric(-RiF %*% matrix(Ss[2:(length(Ss)-1)],nrow=nrow(par$A)))

  s <- sol[,ncol(sol)]
  S <- array(sol[,2:(ncol(sol)-1)],c(length(times),nrow(A),nrow(A)))
  L <- apply(sol,1,SvecToGain)

  L <- array(t(L),c(length(times),ncol(F),nrow(A)))

  return(list(times=times,A=A,F=F,G=G,Q=Q,R=R,P=P,S=S,s=s,L=L))
}

#' Solve the optimal control problem using policy iteration, i.e. find the optimal strategy
#' and value function for the case where the uncontrolled system is given by a subgenerator G0
#' (either due to discounting or due to absorbing boundaries)
#'
#' @param  G0 The sub-generator of the uncontrolled system 
#' @param  G1 A list of (sub-)generators for each control
#' @param  k The running cost 
#' @param  uopt A list of functions returning optional controls as function of 
#' @param  iter.max = 1000 Maximum number of iterations
#' @param  tol = 1e-12 Tolerance for convergence 
#' @param  do.minimize=TRUE
#' @param  do.return.QSD=FALSE Compute and return the quasi-stationary distribution 
#' @return A list containing
#' V: The value function, as a vector with an element for each state
#' u: The optimal controls, as a matrix with a row for each state and a column for each control
#'
#' and, if do.return.QSD==TRUE,
#' qsd.value: The decay rate of the quasi-stationary distribution (decay rate)
#' qsd.vector: The quasi-stationary distribution
#' 
#' @examples
#' ## Controlling a system to the boundary with minimum effort
#' xi <- seq(-2,2,length=101)
#' xc <- as.numeric(cell.centers(xi,c(0,1))$x)
#' dx <- diff(xi)
#' 
#' G0 <- fvade(function(x)-x,function(x)1,xi,'a')
#' Gp <- fvade(function(x)1,function(x)0,xi,'a')
#' Gn <- fvade(function(x)-1,function(x)0,xi,'a')
#' 
#' uopt <- function(dV)pmax(0,-dV)
#' k <- function(u) 1 + 0.5*u[,1]^2 + 0.5*u[,2]^2
#' sol <- PolicyIterationRegular(G0,list(Gp,Gn),k,list(uopt,uopt),do.return.QSD=TRUE)
#' 
#' par(mfrow=c(1,3))
#' plot(xc,sol$V,xlab="x",ylab="Value function",type="l")
#' plot(xc,sol$u[,1]-sol$u[,2],type="l",xlab="x",ylab="Optimal control")
#' plot(xc,sol$qsd.vector/dx,type="l",xlab="x",ylab="QSD",main=sol$qsd.value)
#' @export
PolicyIterationRegular <- function(G0,G1,k,uopt,iter.max = 1000,tol = 1e-12,
                                   do.minimize=TRUE,do.return.QSD=FALSE)
{
    if(!is.list(G1)) G1 <- list(G1)
    if(!is.list(uopt)) uopt <- list(uopt)

    Vold <- (2*do.minimize-1)*Inf

    ## Initial guess on u obtained with a zero value function 
    nu <- length(uopt)
    u <- array(NA,c(nrow(G0),nu))
    for(i in 1:nu) u[,i] <- uopt[[i]](numeric(nrow(G0)))

    iter <- 1
    
    while(TRUE)
    {
        ## Evaluate performance of current strategy
        Gcl <- G0
        for(i in 1:length(G1)) Gcl <- Gcl + Diagonal(x=u[,i]) %*% G1[[i]]
        rhs <- -k(u)
        V <- as.numeric(Matrix::solve(Gcl,rhs))

        ## Determine optimal strategy in response to the current value
        for(i in 1:nu) u[,i] <- uopt[[i]](as.numeric(G1[[i]] %*% V))
        
        if( xor(max(V-Vold)<tol,do.minimize)) break
        if(iter > iter.max) break
        
        Vold <- V
        iter <- iter + 1
    }

    if(do.return.QSD)
    {
        qsd <- QuasiStationaryDistribution(Gcl)
        return(list(V=V,u=u,
                    qsd.value=qsd$value,
                    qsd.vector=qsd$vector))
    }
    
    return(list(V=V,u=u))
}

#' Solve the optimal control problem using policy iteration, i.e. find the optimal strategy
#' and value function for the case where the uncontrolled system is given by a generator G0
#'
#' @param  G0 The generator of the uncontrolled system 
#' @param  G1 A list of generators for each control
#' @param  k The running cost 
#' @param  uopt A list of functions returning optional controls as function of 
#' @param  iter.max = 1000 Maximum number of iterations
#' @param  tol = 1e-12 Tolerance for convergence 
#' @param  do.minimize=TRUE
#' @return A list containing
#' V: The value function, as a vector with an element for each state
#' u: The optimal controls, as a matrix with a row for each state and a column for each control
#' pi: The stationary distribution for the controlled system
#' gamma: The expected running cost in stationarity
#'
#' @examples
#' require(SDEtools)
#' 
#' u <- function(x)0*x
#' D <- function(x) 0*x + 0.25
#' 
#' 
#' xi <- seq(-2,2,length=201)
#' xc <- 0.5*(head(xi,-1) + tail(xi,-1))
#' 
#' G0 <- fvade(u,D,xi,'r')
#' Gp <- fvade(function(x)1,function(x)0,xi,'r')
#' Gn <- fvade(function(x)-1,function(x)0,xi,'r')
#' 
#' uopt <- function(Wp) pmax(0,-Wp)
#' k <- function(u) 0.5*xc^2 + 0.5*u[,1]^2 + 0.5*u[,2]^2
#' 
#' sol <- PolicyIterationSingular(G0,list(Gp,Gn),k,list(uopt,uopt))
#' 
#' par(mfrow=c(1,2))
#' plot(xc,sol$V,type="l",xlab="x",ylab="Value function")
#' plot(function(x)0.5*x^2+min(sol$V),from=-2,to=2,lty="dashed",add=TRUE)
#' plot(xc,sol$u[,1]-sol$u[,2],type="l",xlab="x",ylab="Control")
#' abline(0,-1,lty="dashed")
#' @export
PolicyIterationSingular <- function(G0,G1,k,uopt,iter.max = 1000,tol = 1e-12,do.minimize=TRUE)
{
    if(!is.list(G1)) G1 <- list(G1)
    if(!is.list(uopt)) uopt <- list(uopt)

    ## Initial guess on the optimal performance
    gammaold <- (2*do.minimize-1)*Inf

    ## Initial guess on u obtained with a zero value function 
    nu <- length(uopt)
    u <- array(NA,c(nrow(G0),nu))
    V <- numeric(nrow(G0))

    e <- rep(1,nrow(G0))

    iter <- 1
    
    while(TRUE)
    {
        ## Determine optimal strategy in response to the current value
        for(i in 1:nu) u[,i] <- uopt[[i]](as.numeric(G1[[i]] %*% V))

        ## Construct closed loop generator
        Gcl <- G0
        for(i in 1:length(G1)) Gcl <- Gcl + Diagonal(x=u[,i]) %*% G1[[i]]

        ## Extend with ones right and below; have 0 in the bottom right
        Ge <- rbind(cbind(Gcl,-e),c(e,0))

        ## Solve for the value function and the average performance
        rhs <- c(-k(u),0)
        Vg <- Matrix::solve(Ge,rhs)
        V <- as.numeric(head(Vg,-1))
        gamma <- as.numeric(tail(Vg,1))

        if(xor(gamma-gammaold<tol,do.minimize)) break
        if(iter > iter.max) break

        gammaold <- gamma
        iter <- iter + 1
    }

    pi <- StationaryDistribution(Gcl)
    
    return(list(V=V,u=u,pi=pi,gamma=gamma))
}


R/sdetools.R
#' Simulate a sample path of Brownian motion
#'
#' @name rBM
#' 
#' @param times numeric vector of time points where the Brownian motion should be simulated
#' @param sigma numeric scalar, noise level in the Brownian motion, defaults to 1
#' @param B0 numeric scalar, initial condition (at time=0) for the Brownian motion
#' @param u numeric scalar, drift
#' @return numeric vector, of same length as times, containing the simulated value of the Brownian motion
#' @seealso \code{\link{rvBM}} for a vectorized version
#' @examples
#' times <- 0:10
#' plot(times,rBM(times),type="b",xlab="Time",ylab="Brownian motion")
#' 
#' @export
rBM <- function(times,sigma=1,B0=0,u=0)
{
  dt <- c(times[1],diff(times))
  dB <- rnorm(length(times),mean=u*dt,sd=sigma*sqrt(dt))
  B <- B0+cumsum(dB)
  return(B)
}

#' Simulate a multivariate Brownian motion
#'
#' @name rvBM
#' 
#' @param times numeric vector of time points where the Brownian motion should be simulated
#' @param n numeric scalar, dimension of Brownian motion
#' @param sigma noise level in the Brownian motion. A vector of length n, one for each dimension, or a scalar, in which case the same level is applied to all dimensions.
#' @param B0 initial condition, applicable at time t=0. A vector of length n, one for each dimension, or a scalar, in which case the same initial condition is applied to all dimensions.
#' @param u Drift. An optional numeric n-vector which defaults to a vector of zeros. When supplied, a linear drift (bias) is added to each component.
#' @return a numeric array, n*length(times), each column containing a sample path
#' @examples
#' times <- 0:10
#' VB <- rvBM(times,3)
#' matplot(times,VB,type="b")
#' 
#' @export
rvBM <- function(times,n=1,sigma=rep(1,n),B0=rep(0,n),u=rep(0,n))
{
  if(length(sigma)==1)sigma <- rep(sigma,n)
  if(length(B0)==1)B0 <- rep(B0,n)
  if(length(u)==1)u <- rep(u,n)
  return(sapply(1:n,function(i)rBM(times,sigma=sigma[i],B0=B0[i],u=u[i])))
}

#' Simulate a Brownian bridge
#'
#' @name rBrownianBridge
#' 
#' @param times numeric vector of time points where the Brownian motion should be simulated
#' @param sigma scalar; the noise level in the Brownian motion.
#' @param B0T initial and terminal condition; the fixed value of the Brownian motion at the first and last time points
#' @return a numeric vector, length(times), containing a sample path
#' @examples
#' times <- 10:20
#' BB <- rBrownianBridge(times,c(10,-10),2)
#' plot(times,BB,type="b")
#' 
#' @export
rBrownianBridge <- function(times,B0T=c(0,0),sigma=1)
{
    B <- rBM(times,sigma=sigma)
    n <- length(times)
    t1 <- times[1]
    t2 <- times[n]

    slope <- (B0T[2] - B0T[1] - B[n] + B[1])/(times[n]-times[1])

    BB <- B + B0T[1] - B[1] + slope * (times - times[1])
    return(BB)
}

#' Integrate one sample path of a stochastic process w.r.t. another,
#' returning the Ito integral, the Stratonovich integral, or the
#' "right hand rule".
#'
#' @name stochint
#' 
#' @param f numeric vector containing the integrator
#' @param g numeric vector, same length as f, containing the integrand
#' @param rule character vector indicating rule(s). Valid choices are "l", "r", or "c" for "left", "right", and "center", as well as combinations, e.g. c("l","c"). Optional and defaults to "l".
#' @return A numeric vector, same length as f, giving the "running integral", i.e. the integral as a function of the upper limit.
#' @examples
#' ## Integrating a cosine w.r.t. a sine
#' times <- seq(0,2*pi,length=21)
#' I <- stochint(cos(times),sin(times))
#' Ia <- 0.5*times+0.25*sin(2*times)  # Analytical result
#' matplot(times,I,type="l")
#' lines(times,Ia,col="blue",lwd=2)
#'
#' ## Integration of Brownian motion w.r.t. itself
#' times <- seq(0,10,0.01)
#' BM <- rBM(times)
#' I <- stochint(BM,BM,c("l","c","r"))
#' matplot(times,cbind(I$l,0.5*BM^2-0.5*times),type="l",xlab="Time",ylab="Left integral (Ito)",
#'          main="Integral of B.M. w.r.t itself")
#' matplot(times,cbind(I$r,0.5*BM^2+0.5*times),type="l",xlab="Time",ylab="Right integral",
#'          main="Integral of B.M. w.r.t itself")
#' matplot(times,cbind(I$c,0.5*BM^2),type="l",xlab="Time",ylab="Central integral (Stratonovich)",
#'          main="Integral of B.M. w.r.t itself")
#' 
#' @export
stochint <- function(f,g,rule="l")
{
  n <- length(f)
  dg <- diff(g)
  l <- c(0,cumsum(f[-n]*dg))
  r <- c(0,cumsum(f[-1]*dg))
  c <- 0.5*(l+r)
  return(data.frame(l=l,r=r,c=c)[,rule])
}

#' Discretized quadratic variation of a stochastic process 
#'
#' @name QuadraticVariation
#' 
#' @param X numeric vector containing the process
#' @return A numeric vector, same length as X, giving the discretized quadratic variation as a function of time
#' @examples
#' ## Quadratic Variation of Brownian motion
#' times <- seq(0,10,0.01)
#' B <- rBM(times)
#' plot(times,QuadraticVariation(B))
#'
#' ## Quadratic Variation of an Ito integral
#' G <- cos(B)
#' X <- itointegral(G,B)
#' plot(times,QuadraticVariation(X))
#' lines(times,itointegral(G^2,times))
#' 
#' @export
QuadraticVariation <- function(X)
{
  return(c(0,cumsum(diff(X)^2)))
}

#' Ito integral of a stochastic process w.r.t. another
#'
#' @name itointegral
#' 
#' @param G numeric vector containing the integrator
#' @param B numeric vector, same length as G, containing the integrand
#' @return A numeric vector, same length as G, giving the "running integral", i.e. the Ito integral as a function of the upper limit.
#' @examples
#' ## Integration of Brownian motion w.r.t. itself
#' times <- seq(0,10,0.01)
#' BM <- rBM(times)
#' I <- itointegral(BM,BM)
#' matplot(times,cbind(I,0.5*BM^2-0.5*times),type="l",xlab="Time",ylab="Ito integral",
#'          main="Integral of B.M. w.r.t itself")
#' 
#' @export
itointegral <- function(G,B)
{
  return(c(0,cumsum(G[-length(G)]*diff(B))))
}

#' Cross-variation of two stochastic processes
#'
#' @name CrossVariation
#'
#' @param X numeric vector containing the one process
#' @param Y numeric vector, same length as X, containing the other process
#' @return A numeric vector, same length as X, giving the quadratic cross-varation as function of time
#' @examples
#' ## Quadratic variation of Brownian motion
#' times <- seq(0,10,0.01)
#' B <- rBM(times)
#' plot(times,CrossVariation(B,B),type="l")
#' abline(0,1,lty="dashed")
#'
#' ## Verifying Ito's formula
#' X <- -times + B
#' h <- dhdx <- dh2dx2 <- function(x) exp(x)
#' Y <- h(X)
#' plot(times,Y,type="l")
#'
#' Yi <- h(X[1]) + itointegral(dhdx(X),X) + 0.5*itointegral(dh2dx2(X),CrossVariation(X,X))
#' lines(times,Yi,col="blue")
#' 
#' @export
CrossVariation <- function(X,Y)
{
  return(c(0,cumsum(diff(X)*diff(Y))))
}

#' Heun's method for simulation of a Stratonovich stochastic differential equation dX = f dt + g dB
#'
#' @name heun
#'
#' @param f function f(x) or f(t,x) giving the drift term in the SDE
#' @param g function g(x) or g(t,x) giving the noise term in the SDE
#' @param times numeric vector of time points, increasing
#' @param x0 numeric vector giving the initial condition
#' @param B sample path of (multivariate) Browniań motion. If omitted, a sample is drawn using rvBM
#' @param p Optional projection function that at each time poins projects the state, e.g. to enforce non-negativeness
#' @examples
#' times <- seq(0,10,0.1)
#'
#' # Plot a sample path of the solution of the SDE dX=-X dt + dB
#' plot(times,heun(function(x)-x,function(x)1,times,0)$X,type="l")
#'
#' # Plot a sample path of the solution of the 2D time-varying SDE dX=(AX + FU) dt + G dB
#'
#' A <- array(c(0,-1,1,-0.1),c(2,2))
#' F <- G <- array(c(0,1),c(2,1))
#'
#' f <- function(t,x) A %*% x + F * sin(2*t)
#' g <- function(x) G
#'
#' BM <- rBM(times)
#'
#' plot(times,heun(f,g,times,c(0,0),BM)$X[,1],type="l")
#'
#' # Add solution corresponding to different initial condition, same noise
#'
#' lines(times,euler(f,g,times,c(1,0),BM)$X[,1],type="l",lty="dashed")
#'
#' # Solve geometric Brownian motion as a Stratonovich equation
#' times <- seq(0,10,0.01)
#' BM <- rBM(times)
#' plot(times,heun(function(x)-0.1*x,function(x)x,times,1,BM)$X,type="l",xlab="Time",ylab="X")
#'
#' # Add solution of corresponding Ito equation
#' lines(times,euler(function(x)0.4*x,function(x)x,times,1,BM)$X,lty="dashed")
#'
#' # Testing a case with non-commutative noise
#' BM2 <- rvBM(times,2)
#' matplot(times,heun(function(x)0*x,
#'   function(x)diag(c(1,x[1])),times,c(0,0),BM2)$X,xlab="Time",ylab="X",type="l")
#'
#' # Add solution to the corresponding Ito equation
#' matplot(times,euler(function(x)0*x,
#'   function(x)diag(c(1,x[1])),times,c(0,0),BM2)$X,xlab="Time",ylab="X",type="l",add=TRUE)
#' 
#' @export
heun <- function(f,g,times,x0,B=NULL,p=function(x)x)
{
  nx <- length(x0)
  nt <- length(times)

  ## Check if f is specified as a function(x) only, then convert to a function(t,x)
  if(length(formals(f))==1){
    ff <- function(t,x)f(x)
  }else
  {
    ff <- f
  }

  ## Check if g is specified as a function(x) only, then convert to a function(t,x)
  if(length(formals(g))==1){
    ggg <- function(t,x)g(x)
  }else
  {
    ggg <- g
  }

  ## Find number of dimensions of the Brownian motion. Convert g, if necessary, to something that
  ## returns a nx-by-nB matrix
  g0 <- ggg(times[1],x0)

  if(is.null(dim(g0))) {
      nB <- 1
      gg <- function(t,x)matrix(ggg(t,x),nrow=nx,ncol=1)
  } else {
      nB <- ncol(g0)
      gg<-ggg
  }

  ## If the sample path of Brownian motion is not specified, then simulate it
  if(is.null(B)){
    B <- rvBM(times,nB)
  }

  if(!is.matrix(B)) B <- matrix(B,nrow=nt,ncol=nB)

  dB <- apply(B,2,diff)

  X <- array(NA,c(nt,nx))
  X[1,] <- x0

  dt <- diff(times)

  for(i in 1:(nt-1))
  {
    ## Euler predictor
    fX <- as.numeric(ff(times[i],X[i,]))
    gX <- gg(times[i],X[i,])
    Y <- as.numeric(p(X[i,] + fX*dt[i] + gX %*% as.numeric(dB[i,])))

    ## Corrector
    fY <- as.numeric(ff(times[i+1],Y))
    gY <- gg(times[i+1],Y)
    X[i+1,] <- as.numeric(p(X[i,] + 0.5*(fX+fY)*dt[i] + 0.5*(gX+gY) %*% as.numeric(dB[i,])))
  }

  colnames(X) <- names(x0)
  
  return(list(times=times,X=X))
}


#' Euler simulation of an Ito stochastic differential equation dX = f dt + g dB
#'
#' @name euler
#'
#' @param f function f(x) or f(t,x) giving the drift term in the SDE
#' @param g function g(x) or g(t,x) giving the noise term in the SDE
#' @param times numeric vector of time points, increasing
#' @param x0 numeric vector giving the initial condition
#' @param B sample path of (multivariate) Browniań motion. If omitted, a sample is drawn using rvBM
#' @param p Optional projection function that at each time poins projects the state, e.g. to enforce non-negativeness
#' @examples
#' times <- seq(0,10,0.1)
#'
#' # Plot a sample path of the solution of the SDE dX=-X dt + dB
#' plot(times,euler(function(x)-x,function(x)1,times,0)$X,type="l")
#'
#' # Plot a sample path of the solution of the 2D time-varying SDE dX=(AX + FU) dt + G dB
#'
#' A <- array(c(0,-1,1,-0.1),c(2,2))
#' F <- G <- array(c(0,1),c(2,1))
#'
#' f <- function(t,x) A %*% x + F * sin(2*t)
#' g <- function(x) G
#'
#' BM <- rBM(times)
#'
#' plot(times,euler(f,g,times,c(0,0),BM)$X[,1],type="l")
#'
#' # Add solution corresponding to different initial condition, same noise
#'
#' lines(times,euler(f,g,times,c(1,0),BM)$X[,1],type="l",lty="dashed")
#' 
#' @export
euler <- function(f,g,times,x0,B=NULL,p=function(x)x)
{
  nx <- length(x0)
  nt <- length(times)

  ## Check if f is specified as a function(x) only, then convert to a function(t,x)
  if(length(formals(f))==1){
    ff <- function(t,x)f(x)
  }else
  {
    ff <- f
  }

  ## Check if g is specified as a function(x) only, then convert to a function(t,x)
  if(length(formals(g))==1){
    ggg <- function(t,x)g(x)
  }else
  {
    ggg <- g
  }

  ## Find number of dimensions of the Brownian motion. Convert g, if necessary, to something that
  ## returns a nx-by-nB matrix
  g0 <- ggg(times[1],x0)

  if(is.null(dim(g0))) {
      nB <- 1
      gg <- function(t,x)matrix(ggg(t,x),nrow=nx,ncol=1)
  } else {
      nB <- ncol(g0)
      gg<-ggg
  }

  ## If the sample path of Brownian motion is not specified, then simulate it
  if(is.null(B)){
    B <- rvBM(times,nB)
  }

  if(!is.matrix(B)) B <- matrix(B,nrow=nt,ncol=nB)

  dB <- apply(B,2,diff)

  X <- array(NA,c(nt,nx))
  colnames(X) <- names(x0)
  X[1,] <- x0

  dt <- diff(times)

  for(i in 1:(nt-1))
  {
    X[i+1,] <- p( X[i,] + ff(times[i],X[i,])*dt[i] + gg(times[i],X[i,]) %*% as.numeric(dB[i,]) )
  }

  colnames(X) <- names(x0)

  return(list(times=times,X=X))
}

#' Transition probabilities in a linear SDE dX = A*X*dt + u*dt + G*dB
#'
#'
#' @name dLinSDE
#'
#' @param A system matrix, quadratic numeric array or matrix
#' @param G noise matrix, same number of rows as A
#' @param t time, numeric scalar
#' @param x0 Initial state
#' @param u forcing. A numeric vector or matrix; see details
#' @param S0 Initial variance
#' @return A list containing St, the variance of the state at time t, and, depending on input arguments
#'
#'   eAt, the matrix that maps X0 to Xt, i.e. expm(A*t), provided x0==NULL
#'
#'   EX, the expectation of the state at time t, provided x0 is not NULL
#' @details
#' When returning EX, the expectation at time t, the input u can be specified as follows:
#'
#' * NULL, in which case it is assumed to be absent from the equation
#'
#' * a numeric vector with an element for each row in A. In this case u is assumed to be constant
#'
#' * a numeric array or matrix with same number of rows as A and two columns. In this case u is assumed
#'   to vary affinely in time, equaling the first column at time 0 and the second column at time t.
#'
#' The computations are not optimized for large systems, since they rely on the vector form of the matrix equations and do not use sparsity.
#' @examples
#' # A scalar equation with no input
#' (dLinSDE(-1,1,1))
#' # A scalar equation with constant input, starting in steady-state
#' (dLinSDE(A=-1,G=1,t=3,x0=1,u=1,S0=0.5))
#' 
#' @export
dLinSDE <- function(A,G,t,x0=NULL,u=NULL,S0=0*A)
{
    G <- as.matrix(G)
    A <- as.matrix(A)

    GG <- G %*% t(G)

    nx <- nrow(A)
    nu <- ncol(G)

    eAt <- as.matrix(Matrix::expm(A*t))

    St <- tryCatch(
        expr = {
            Sinf <- lyap(A,GG)
            St <- Sinf - eAt %*% (Sinf - S0) %*% t(eAt)            
        },
        error = function(e){
            I <- Matrix::Diagonal(n=nx,x=1)
            M <- A %x% I + I %x% A
            P <- rbind(array(0,c(nx^2,2*nx^2)),cbind(diag(rep(1,nx^2)),M))
            gs <- Matrix::expm(P*t) %*% c(as.numeric(GG),rep(0,nx^2))
            St <-eAt %*% S0 %*% t(eAt) + matrix(gs[nx^2+(1:(nx^2))],nrow=nx)
        })

    if(is.null(x0)){
      return(list(eAt = eAt,St = St))
    }

    if(is.null(u)){
      return(list(EX=eAt %*% x0,St=St))
    }

    ## Zero order hold
    if(length(u)==nu){
      AA <- rbind(cbind(A,G),array(0,c(nu,nu+nx)))
      uEx <- Matrix::expm(AA*t) %*% c(x0,u)
      return(list(EX=uEx[nx+(1:nx)],St=St))
    }

    u = as.matrix(u)

    ## First order hold
    if(all(dim(u)==c(nu,2))){
      O <- array(0,c(nx,nx))
      AAA <- rbind(numeric(2*nu+nx),
                   cbind(diag(nrow=nu),array(0,c(nu,nu+nx))),
                   cbind(arrayu(0,c(nx,nu)),G,A))

      duuEx <- Matrix::expm(AAA*t) %*% c((u[,2]-u[,1])/t,u[,1],x0)
      return(list(EX=duuEx[2*nx+(1:nx)],St=St))
    }
}

#' Algebraic Lyapunov equation A*X+X*t(A)+Q=0
#'
#' @name lyap
#'
#' @param A A quadratic matrix without eigenvalues on the imaginary axis
#' @param Q A symmetric matix of same dimension as A
#' @return X A symmetric matrix of same dimension as A
#'
#' @details
#' If A is asymptotically stable, Q is positive semidefinite and the pair (A,Q) is controllable, then X will be positive definite. Several similar results exist.
#' The implementation uses vectorization and kronecker products and does not employ sparsity, so is only suitable for small systems. 
#'
#' @examples
#' # A scalar example
#' (lyap(-1,1))
#' # A harmonic oscillator
#' (lyap(array(c(0,-1,1,-0.1),c(2,2)),diag(c(0,1))))
#'
#' @export
lyap <- function(A,Q) 
{
    A <- as.matrix(A)
    I <- diag(rep(1,nrow(A)))
    P <- kronecker(I,A)+kronecker(A,I)
    X <- -solve(P,as.numeric(Q))
    return(matrix(X,nrow=nrow(A)))
}

